network/crl.go

```diff
@@ -0,0 +1,503 @@
+
+var (
+	crlFromCertValValues = map[string]int{
+		CrlFromCertUseStr:    crlFromCertUse,
+		CrlFromCertTrustStr:  crlFromCertTrust,
+		CrlFromCertPreferStr: crlFromCertPrefer,
+		CrlFromCertIgnoreStr: crlFromCertIgnore,
+	}
+)
+
+const (
+	crlFromCertUse int = iota
+	crlFromCertTrust
+	crlFromCertPrefer
+	crlFromCertIgnore
+)
+
+const (
+	// Max int value, because cache implementation uses `int` for setting capacity
+	CrlCacheSizeMax = 2147483647
+	CrlCacheTimeMax = 300
+)
+
+// CRLConfig contains configuration related to certificate validation using CRL
+type CRLConfig struct {
+	url                      string
+	fromCert                 int // crlFromCert*
+	checkOnlyLeafCertificate bool
+	cacheSize                uint
+	cacheTime                time.Duration
+	ClientAuthType           tls.ClientAuthType
+}
+
+// NewCRLConfig creates new CRLConfig
+func NewCRLConfig(url, fromCert string, checkOnlyLeafCertificate bool, cacheSize, cacheTime uint) (*CRLConfig, error) {
+	fromCertVal, ok := crlFromCertValValues[fromCert]
+	if !ok {
+		return nil, ErrInvalidConfigCRLFromCert
+	}
+
+	if cacheSize > CrlCacheSizeMax {
+		return nil, ErrInvalidConfigCRLCacheSize
+	}
+
+	if cacheTime > CrlCacheTimeMax {
+		return nil, ErrInvalidConfigCRLCacheTime
+	}
+
+	if url != "" {
+		_, err := url_.Parse(url)
+		if err != nil {
+			return nil, err
+		}
+
+		// Since this is CRL configuration alone, we don't have access to cache yet;
+		// so let's just download the CRL and forget about it
+		crlClient := NewDefaultCRLClient()
+		_, err = crlClient.Fetch(url, true)
+		if err != nil {
+			log.WithError(err).WithField("url", url).Warnln("CRL: Cannot fetch configured URL")
+		}
+	}
+
+	return &CRLConfig{
+		url:                      url,
+		fromCert:                 fromCertVal,
+		checkOnlyLeafCertificate: checkOnlyLeafCertificate,
+		cacheSize:                cacheSize,
+		cacheTime:                time.Second * time.Duration(cacheTime),
+		ClientAuthType:           tls.RequireAndVerifyClientCert,
+	}, nil
+}
+
+// UseCRL returns true if verification via CRL is enabled
+func (c *CRLConfig) UseCRL() bool {
+	if c == nil {
+		return false
+	}
+	return c.url != "" || c.fromCert != crlFromCertIgnore
+}
+
+func (c *CRLConfig) isCachingEnabled() bool {
+	return c.cacheTime > 0 && c.cacheSize > 0
+}
+
+// CRLClient is used to fetch CRL from some URL
+type CRLClient interface {
+	// Fetch fetches CRL from passed URL (can be either http:// or file://),
+	// allowLocal controls whether file:// will be handled (should not be allowed for URLs from certificates)
+	Fetch(url string, allowLocal bool) ([]byte, error)
+}
+
+// DefaultCRLClient is a default implementation of CRLClient
+// (as opposed to stub ones used in tests)
+type DefaultCRLClient struct {
+	httpClient *http.Client
+}
+
+// NewDefaultCRLClient creates new DefaultCRLClient
+func NewDefaultCRLClient() DefaultCRLClient {
+	return DefaultCRLClient{httpClient: &http.Client{
+		Timeout: time.Second * time.Duration(20),
```

Comment:
better to move this value `time.Second * time.Duration(20)` to constant value. it will allow to re-use it in tests to verify that timeout works as expected and allow to avoid copy-paste. Plus if some other part of acra will want to keep in mind such timeouts, it may access to this value if need (for example to sum or aggregate all timeouts in some chain of calls)

Reply:
Extracted HTTP client default timeouts to constants (`OcspHttpClientDefaultTimeout` and `CrlHttpClientDefaultTimeout`)